src/Stratis.Features.FederatedPeg/SourceChain/MaturedBlocksProvider.cs

```diff
@@ -60,30 +67,52 @@ public async Task<Result<List<MaturedBlockDepositsModel>>> GetMaturedDepositsAsy
             int maxTimeCollectionCanTakeMs = RestApiClientBase.TimeoutMs / 2;
             var cancellation = new CancellationTokenSource(maxTimeCollectionCanTakeMs);
 
-            for (int i = blockHeight; (i <= matureTipHeight) && (i < blockHeight + maxBlocks); i++)
+            int maxBlockHeight = Math.Min(matureTipHeight, blockHeight + maxBlocks - 1);
+
+            var headers = new List<ChainedHeader>();
+            ChainedHeader header = consensusTip.GetAncestor(maxBlockHeight);
+            for (int i = maxBlockHeight; i >= blockHeight; i--)
             {
-                ChainedHeader currentHeader = consensusTip.GetAncestor(i);
+                headers.Add(header);
+                header = header.Previous;
+            }
 
-                ChainedHeaderBlock block = this.consensusManager.GetBlockData(currentHeader.HashBlock);
+            headers.Reverse();
 
-                if (block?.Block?.Transactions == null)
-                {
-                    // Report unexpected results from consenus manager.
-                    this.logger.LogWarning("Stop matured blocks collection due to consensus manager integrity failure. Send what we've collected.");
-                    break;
-                }
+            int numDeposits = 0;
 
-                MaturedBlockDepositsModel maturedBlockDeposits = this.depositExtractor.ExtractBlockDeposits(block);
+            for (int ndx = 0; ndx < headers.Count; ndx += 100)
+            {
+                List<ChainedHeader> currentHeaders = headers.GetRange(ndx, Math.Min(100, headers.Count - ndx));
 
-                if (maturedBlockDeposits == null)
-                    throw new InvalidOperationException($"Unable to get deposits for block at height {currentHeader.Height}");
+                List<uint256> hashes = currentHeaders.Select(h => h.HashBlock).ToList();
 
-                maturedBlocks.Add(maturedBlockDeposits);
+                ChainedHeaderBlock[] blocks = this.consensusManager.GetBlockData(hashes);
```

Comment:
AS I said you run the risk to get blocks not on best chain here and alternative approach would be something like this

```
var blockHeaders = this.chainIndexer.GetBlockHeaders(hashes)

ChainedHeaderBlock[] blocks = this.blockStore.GetBlocks(blockHeaders )
```

Reply:
The following will only require a new method on `ChainIndexer` so perhaps this can be the first iteration in a follow-up PR:
```
                List<ChainedHeader> blockHeaders = this.chainIndexer.EnumerateRange(height, Math.Min(100, maxBlockHeight - height + 1)).ToList();

                List<uint256> hashes = blockHeaders.Select(h => h.HashBlock).ToList();

                ChainedHeaderBlock[] blocks = this.consensusManager.GetBlockData(hashes);
```